### ISSUE TITLE
how to implement remote validation?

### ISSUE THREAD

**Issue by hxlniada** [ghiscoding/angular-validation#34]
how to implement remote validation?

**Comment by ghiscoding** [comment]
Oh yes I forgot to add it to the Wiki when I created them. See this [Wiki - Remove-Validator-from-Element](https://github.com/ghiscoding/angular-validation/wiki/Remove-Validator-from-Element) but this only works if you defined your validation through the Service, it will not work if your elements were defined through the Directive.

**Comment by ghiscoding** [comment]
I revisited the code and found a way to make this available from both the Service and Directive. It will be available in the next version, should done by the next weekend. 

I will leave this issue open until I push the next version with the fix. 
Thanks for the feedback.

**Comment by hxlniada** [comment]
This validation tool is very convenient.

**Comment by hxlniada** [comment]
But I think you misunderstand my word... It's  "remote" not "remove"...
I think the custom validator can return a promise or something else .

**Comment by ghiscoding** [comment]
Oh wow ok 1 character could cause a completely different understanding haha

I'm not sure about your question about remote then... did you look at the [Service](https://github.com/ghiscoding/angular-validation/wiki/Working-Service-Examples)? With the Service you can define your validation through the Controller, unless I misunderstood what you want to do, then please give me some examples of what you are trying to accomplish.

**Comment by hxlniada** [comment]
What I want is something like:
```html
<input validation="required | custom:customValidator | ...
```
the customValidator is someting like
```javascript
$scope.customValidator = function(value) {
    var defer = $q.defer();
    $http.get('validationLink').success(function(data) {
        if (data.code === 200) {
            defer.resolve(true);
        } else {
            defer.resolve(errorMsg)
        }
    }).error(function() {
        defer.reject();
    });
}
```

**Comment by hxlniada** [comment]
Sorry for my poor english, I wish I had explained it clearly...

**Comment by ghiscoding** [comment]
Hello, your english is understandable, no problem ;)
You can do what you are describing with the [Regex Validator](https://github.com/ghiscoding/angular-validation/wiki/Regular-Expressions-(Regex)) and you can see it in action at the [Example #6 here](https://github.com/ghiscoding/angular-validation/wiki/Working-Directive-Examples)

This is the code
```html
!-- example 6 - with Regular Expression (Date Code of YYWW) -->
<label for="input5">Multiple Validations + Custom Regex of Date Code (YYWW)</label>
<input type="text" name="input5" ng-model="form1.input5"
        validation="exact_len:4|regex:YYWW:=^(0[9]|1[0-9]|2[0-9]|3[0-9])(5[0-2]|[0-4][0-9])$:regex|required|integer"  />
```
Make sure to read how to use it [Regex Validator](https://github.com/ghiscoding/angular-validation/wiki/Regular-Expressions-(Regex))

**Comment by hxlniada** [comment]
Thank you..., I had read through the wiki and I know the custom RGEX validation, But i don't think remote validation can be accomplished via REGEX...
This problem is resolved, thank you again

**Comment by ghiscoding** [comment]
Ah ok I think I understand what you mean now, you want to call an external possibly AJAX function, that might be doable. I will have to think about how to do that though. If you know how to do it, then please share some code, else it might take a while before I implement it.

**Comment by hxlniada** [comment]
@ghiscoding I will have a try after about three days. For now I have too much things to do...

**Comment by geemang2000** [comment]
:+1:   I have a project that has a number of remote validation calls.

**Comment by hxlniada** [comment]
```javascript
// is promise
if (angular.isFunction(custom.then)) {
    custom.then(function (result) {
        if (result === true) {
            ...$valid;
        } else {
             $invalid;// result maybe the error message
        }
    })
}

**Comment by ghiscoding** [comment]
I think that I got something working, still need to do a few more tests but it seems to be working as expected.

So for that work, I expect a `promise`, so I made a simple promise in my test controller and return a false value after 1sec. There is 2 ways of resolving a result, it's probably better with an object (the 2nd one).
```javascript
$scope.customRemoteValidationCall = function() {
    var deferred = $q.defer();
    setTimeout(function() {
      // you can return a boolean for isValid
      deferred.resolve(false);

      // or you can return an object as { isValid: bool, message: msg }
      deferred.resolve({ isValid: false, message: 'Returned error from promise'});
    }, 1000);

    return deferred.promise;
}
```
then calling it inside the DOM with the Directive would be like below, also note that you can use this validator in combination with any other validators as it's shown below and you can also use the alternate text but it might not be best to use since defining an `alt=` would display the error message right away while resolving through the promise as a result object would only display it at the end. And to make it a bit more clear, all other validator error messages are showing right away but when the remote call is resolving it will concatenate it to the error message.
```html
<!-- defining the remote function call by using `remote:` -->
<input type="text" class="form-control" name="input3" ng-model="input3" 
    validation="alpha|min_len:2|remote:customRemoteValidationCall:alt=Remote error|required" />

<!-- using alternate text but is probably not suggested -->
<input type="text" class="form-control" name="input3" ng-model="input3" 
    validation="alpha|min_len:2|remote:customRemoteValidationCall:alt=Remote error|required" />
```

I'd like to have some feedback before pushing it, even though I still need couple more tests to do, it's looking good... feedback please!

**Comment by ghiscoding** [comment]
@hxlniada Feature added in v1.3.25, you can see the [Wiki - Remote Validation](https://github.com/ghiscoding/angular-validation/wiki/Remote-Validation) for the documentation. 

@geemang2000 if you like my open source Angular-Validation, make sure to add it to your favorite by clicking on the Star on top ;)

Thanks

**Comment by hxlniada** [comment]
Your are the man

**Comment by ghiscoding** [comment]
So I guess you like it :+1: 
It was a great idea, I never thought of it before ;)

Just tell everyone about my nice project... Have a nice one :)

**Comment by ghiscoding** [comment]
As I was starting to use Remote Validation for my own project I found some performance issue in the sense that it was sending a lot more calls to the server than I expected and I found out it was a bug with onBlur that I had since the beginning but never noticed it before. The bug found with onBlur which would run as much validations as there was characters inside the input when onBlur was called (abcdef => would make 6 validations) and this become costly and excessive with Remote Validation to send so many calls to the server. 

So this mentioned bug was fixed in version `1.3.27` and I would strongly recommend you to update especially if you do use the Remote Validation. I also added the possibility to use the "Controller As" alias (As vm or anything else), so you can now write `validation="remote:vm.customRemoteValidation|required"`.

Enjoy it :)
